### PR TITLE
Integrating sync_wait & sync_wait_with_variant

### PR DESCRIPTION
**Pull request by hcq9102**
## Proposed Changes:
Ref: [P2300R5](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2022/p2300r5.html#example-schedulers)
Integrate sync_wait and sync_wait_with_variant.

## Checklist

-  I have added a new feature and have added tests to go along with it.

### CODE REVIEW

**Comment by hkaiser** [comment]
@hcq9102 do we still need to merge #5956 before merging this?

**Comment by hcq9102** [comment]
> @hcq9102 do we still need to merge #5956 before merging this?

I think we don't need #5956 anymore.
But I would like to close #5956 after this PR is merged.

**Comment by hkaiser** [comment]
@hcq9102 this should be fine now. Please note that I have also rebased (and force-pushed) this PR onto master.

File: libs/core/execution/include/hpx/execution/algorithms/sync_wait.hpp [hunk 1]
```diff
@@ -1,5 +1,6 @@
 //  Copyright (c) 2020 ETH Zurich
 //  Copyright (c) 2022 Hartmut Kaiser
+//  Copyright (c) Chuanqiu He
```
Comment (hkaiser) [inline libs/core/execution/include/hpx/execution/algorithms/sync_wait.hpp:3]: Please add a year to the copyright

**Review by hkaiser — APPROVED** [review]
LGTM, thanks!

**Comment by hkaiser** [comment]
bors merge

**Pull request merged**